### PR TITLE
Feature: Add Notional V2 Connector

### PR DESCRIPTION
**Pull request by jeffywu**
- [x] Review and approve connector @pradyuman-verma @thrilok209 
- [x] Deploy and comment addresses.
- [x] Add to dsa-connect https://github.com/Instadapp/dsa-connect/pull/162
- [ ] Add to docs
- [ ] Build and deploy resolver

### CODE REVIEW

**Comment by pradyuman-verma** [comment]
@jeffywu Are there any test cases for the connector?
If not, can you add them!

**Comment by jeffywu** [comment]
@pradyuman-verma we wrote the tests in brownie so @weitianjie2000 will have to port them over to hardhat and add them here.

**Comment by pradyuman-verma** [comment]
> @pradyuman-verma, we wrote the tests in brownie, so @weitianjie2000 will have to port them over to hardhat and add them here.

Sure, Convert them to typescript, and let us know!

**Comment by ghost** [comment]
@pradyuman-verma Hi, I've migrated all of the tests. Please let me know if you need anything else. Thanks!

File: contracts/mainnet/connectors/notional/SafeInt256.sol [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    function neg(int256 x) internal pure returns (int256 y) {
+        return mul(-1, x);
+    }
+
+    function toInt(uint256 x) internal pure returns (int256 y) {
+        require(x <= uint256(type(int256).max));
+        return int256(x);
+    }
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/SafeInt256.sol]: I think we don't have to use this safe library, and we do have the same library in `~/contracts/mainnet/common/math.sol`, which has all this functionality.
If something in it is missing, you can add it in there.

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,272 @@
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]: use this in here https://github.com/Instadapp/dsa-connectors/blob/main/contracts/mainnet/common/math.sol

**Comment by pradyuman-verma** [comment]
@jeffywu The test cases are in `js,` We use `ts.` Can you migrate them to typescript?

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 2 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: done

File: contracts/mainnet/connectors/notional/SafeInt256.sol (hunk 1 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/SafeInt256.sol]: I removed this file and added toUint to DSMath

**Comment by ghost** [comment]
@pradyuman-verma tests are in typescript now

**Comment by ghost** [comment]
InstadappTests

File: contracts/mainnet/connectors/notional/main.sol [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+import {Helpers} from "./helpers.sol";
+import {Events} from "./events.sol";
+import {DepositActionType, BalanceActionWithTrades, BalanceAction} from "./interface.sol";
+import {TokenInterface} from "../../common/interfaces.sol";
+
+abstract contract NotionalResolver is Events, Helpers {
+    /**
+     * @notice Deposit collateral into Notional, this should only be used for reducing risk of
+     * liquidation. 
+    *  @dev Deposits into Notional are not earning fixed rates, they are earning the cToken
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/main.sol]:
```suggestion
     * liquidation. 
     *  @dev Deposits into Notional are not earning fixed rates, they are earning the cToken
```

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import {Basic} from "../../common/basic.sol";
+import {DSMath} from "../../common/math.sol";
+import {TokenInterface} from "../../common/interfaces.sol";
+
+abstract contract Helpers is DSMath, Basic {
+    uint8 internal constant LEND_TRADE = 0;
+    uint8 internal constant BORROW_TRADE = 1;
+    uint256 internal constant INTERNAL_TOKEN_PRECISION = 1e8;
+    uint256 internal constant ETH_CURRENCY_ID = 1;
+    uint256 internal constant MAX_DEPOSIT = uint256(-1);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]:
```suggestion
    uint256 internal constant MAX_DEPOSIT = type(uint256).max;
```

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 5]
```diff
@@ -0,0 +1,286 @@
+pragma solidity ^0.7.6;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol:2]: Can you add SPDX license identifier here

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        return toUint(nTokenBalance);
+    }
+
+    function getNTokenRedeemAmount(
+        uint16 currencyId,
+        uint96 _tokensToRedeem,
+        uint256 getId
+    ) internal returns (uint96 tokensToRedeem) {
+        tokensToRedeem = uint96(getUint(getId, _tokensToRedeem));
+        if (tokensToRedeem == uint96(-1)) {
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]: Use this, if available in version 0.7.6!
```suggestion
        if (tokensToRedeem == type(uint96).max) {
```

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        ) = notional.getAccountBalance(currencyId, address(this));
+        return toUint(nTokenBalance);
+    }
+
+    function getNTokenRedeemAmount(
+        uint16 currencyId,
+        uint96 _tokensToRedeem,
+        uint256 getId
+    ) internal returns (uint96 tokensToRedeem) {
+        tokensToRedeem = uint96(getUint(getId, _tokensToRedeem));
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]: Is there any possibility of uint overflow?

File: contracts/mainnet/connectors/notional/main.sol [hunk 8]
```diff
@@ -0,0 +1,671 @@
...
+            currencyId,
+            useUnderlying,
+            depositAmount
+        );
+
+        notional.batchBalanceAction{value: msgValue}(address(this), action);
+
+        uint256 nTokenBalanceChange = sub(getNTokenBalance(currencyId), nTokenBefore);
+
+        if (setId != 0) {
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/main.sol]: Why this? Can't we pass setId as 0?

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        (/* assetToken */, Token memory underlyingToken) = notional.getCurrency(currencyId);
+        return underlyingToken.tokenAddress;
+    }
+
+    /// @notice Returns the address of the asset token for a given currency id
+    function getAssetToken(uint16 currencyId) internal view returns (address) {
+        // prettier-ignore
+        (Token memory assetToken, /* underlyingToken */) = notional.getCurrency(currencyId);
+        return assetToken.tokenAddress;
+    }
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]: Can we merge these both into a single function?

File: contracts/mainnet/connectors/notional/helpers.sol [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        returns (uint256)
+    {
+        // prettier-ignore
+        (
+            /* int256 cashBalance */,
+            int256 nTokenBalance,
+            /* int256 lastClaimTime */
+        ) = notional.getAccountBalance(currencyId, address(this));
+        return toUint(nTokenBalance);
+    }
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/helpers.sol]: These both too, as they perform the same actions but returns different thing

File: contracts/mainnet/connectors/notional/main.sol [hunk 11]
```diff
@@ -0,0 +1,671 @@
...
+     * lending rate. In order to lend at fixed rates use `depositAndLend`
+     * @param currencyId notional defined currency id to deposit
+     * @param useUnderlying if true, will accept a deposit in the underlying currency (i.e DAI), if false
+     * will use the asset currency (i.e. cDAI)
+     * @param depositAmount amount of tokens to deposit
+     * @param getId id of depositAmount
+     * @param setId id to set the value of notional cash deposit increase (denominated in asset cash, i.e. cDAI)
+     */
+    function depositCollateral(
+        uint16 currencyId,
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/main.sol]: Is there any API endpoint or docs which mentions all currencyID?

File: contracts/mainnet/connectors/notional/main.sol [hunk 12]
```diff
@@ -0,0 +1,671 @@
...
+            amountToWithdraw,
+            redeemToUnderlying
+        );
+    }
+
+    /**
+     * @notice Redeems nTokens and uses the cash to repay a borrow.
+     * @dev When specifying fCashAmount be sure to calculate it such that the account
+     * has enough cash after redeeming nTokens to pay down the debt. This can be done
+     * off-chain using the Notional SDK.
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/notional/main.sol]: Can we do this on-chain, too, inside the contract itself?

**Comment by pradyuman-verma** [comment]
@jeffywu Can you add the latest screenshot of the tests?
Also, Can you resolve the conversations on which you have worked on?

**Comment by ghost** [comment]
new test run,

InstadappTests

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 4 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: done

File: contracts/mainnet/connectors/notional/main.sol (hunk 3 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/main.sol]: done

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 5 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol:2]: done

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 6 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: done

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 7 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: added conversion functions and overflow checks

File: contracts/mainnet/connectors/notional/main.sol (hunk 8 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/main.sol]: This allows the caller to save some gas by setting setId to zero if the get/set mechanism is not used.

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 9 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: done

File: contracts/mainnet/connectors/notional/helpers.sol (hunk 10 above)
Comment (ghost) [reply contracts/mainnet/connectors/notional/helpers.sol]: done

**Comment by pradyuman-verma** [comment]
@weitianjie2000 Thanks.
Can you resolve the conversations on which you have worked?

File: contracts/mainnet/connectors/notional/main.sol (hunk 11 above)
Comment (jeffywu) [reply contracts/mainnet/connectors/notional/main.sol]: Currency ID is autoincrementing from 1 to N and maps to a currency in the system. Currently: ETH is 1, DAI is 2, USDC is 3, WBTC is 4. You can get more details about each currency id from the subgraph:
https://thegraph.com/hosted-service/subgraph/notional-finance/mainnet-v2

Once set, the currency that is mapped to at a particular ID won't change.

File: contracts/mainnet/connectors/notional/main.sol (hunk 12 above)
Comment (jeffywu) [reply contracts/mainnet/connectors/notional/main.sol]: The calculation is not straightforward due to slippage across multiple markets, it's not easy to implement this is in solidity. Note that this method is here for convenience, a user could achieve the same results in two steps without using the off-chain SDK (first redeem, then repay).

File: contracts/mainnet/connectors/notional/main.sol [hunk 13]
```diff
@@ -0,0 +1,671 @@
...
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        tokensToRedeem = getNTokenRedeemAmount(
+            currencyId,
+            tokensToRedeem,
+            getId
+        );
+
+        int256 _assetCashChange = notional.nTokenRedeem(
```
Comment (jeffywu) [inline contracts/mainnet/connectors/notional/main.sol]: @weitianjie2000 update this to include acceptResiduals from the V2.1 upgrade

File: contracts/mainnet/connectors/notional/main.sol (hunk 8 above)
Comment (pradyuman-verma) [reply contracts/mainnet/connectors/notional/main.sol]: @thrilok209 Is this valid?

File: contracts/mainnet/connectors/notional/main.sol (hunk 12 above)
Comment (pradyuman-verma) [reply contracts/mainnet/connectors/notional/main.sol]: Alright, got it

File: contracts/mainnet/connectors/notional/main.sol (hunk 11 above)
Comment (pradyuman-verma) [reply contracts/mainnet/connectors/notional/main.sol]: Alright, thanks

**Comment by pradyuman-verma** [comment]
@jeffywu @weitianjie2000 We can proceed towards deployment.

**Comment by jeffywu** [comment]
> 

@pradyuman-verma, @weitianjie2000 is going to add another method here to allow users to Stake NOTE tokens (we just launched this earlier this week) so I think that will be great for users.

What's the next step for deployment? Do you want us to deploy and verify this contract and send you back an address?

**Comment by pradyuman-verma** [comment]
> > 
> 
> @pradyuman-verma, @weitianjie2000 will add another method here to allow users to Stake NOTE tokens (we just launched this earlier this week), so I think that will be great for users.
> 
> What's the next step for deployment? Do you want us to deploy and verify this contract and send you back an address?

 Let's add the `Stake NOTE tokens` feature, deploy the contract, and comment the addresses here. @jeffywu

**Comment by ghost** [comment]
@pradyuman-verma I've added the staking functions and tests + screenshot for the new test run. Would you like to take another look before I deploy the connector?

InstadappTests

**Comment by pradyuman-verma** [comment]
> @pradyuman-verma I've added the staking functions and tests + screenshot for the new test run. Would you like to take another look before I deploy the connector?
> 
> InstadappTests

Yes, I will review it and let you know if any changes are required.

**Review by pradyuman-verma — APPROVED** [review]

**Comment by pradyuman-verma** [comment]
@jeffywu @weitianjie2000 We can deploy it.

**Review by pradyuman-verma — APPROVED** [review]

**Comment by ghost** [comment]
@pradyuman-verma @jeffywu connector is deployed and verified

**Name** `NOTIONAL-V2-A`
**Addr** [0x0b7675e19d71a0c8953444f7f9913b8448186441](https://etherscan.io/address/0x0b7675e19d71a0c8953444f7f9913b8448186441#code)
**tx** [here](https://etherscan.io/tx/0x1e80b0dda285fe9af8637e6b7189e514beae70bf6acdc37ee1e950bfcf3b2c17#eventlog)

**Pull request merged**